Jimix/23 pushpull

jimix opened the pull request:
Resolves #23
@mkistler: please ACK when you get a chance

jimix committed:
Use CloudantSync interfaces for pushing and pulling

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Allow arbitrary number of secondary remotes

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix committed:
Use new push/pull interfaces in tests

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

mkistler commented on Classes/common/CDTIncrementalStore.h:
```diff
@@ -129,4 +59,34 @@ typedef NS_ENUM(NSInteger, CDTIncrementalStoreErrors) {
+ * Create a CDTReplicator object set up to replicate changes from a remote database to the
+ * local datastore.
+ *
+ *  @param remoteURL the remote server URL to which the data is replicated.
```
"to which" should be "from which"

jimix replied:
ack

mkistler commented on Classes/common/CDTIncrementalStore.m:
```diff
@@ -1571,75 +1370,73 @@ - (void)setMetadata:(NSDictionary *)metadata
+        CDTLogError(CDTREPLICATION_LOG_CONTEXT, @"%@: %@: Could not create push replication object",
+                    CDTISType, clean);
+        return nil;
+    }
```
Seems like we would not need to recreate the repFactory and pushRep every time this method is called.  We could cache these in local properties and reuse them.

jimix replied:
I think it is possible that `repFactory` could be reused, I will check and I think we could just create it up front.
However, the `(push|pull)Rep` objects are actually single use, so we need to create them on demand.

jimix replied:
I have confirmed that `repFactory` can be reused.

mkistler commented on doc/coredata.md:
````diff
@@ -99,49 +99,30 @@ NSURL *linkURL = [NSURL URLWithString:databaseURI];
-if (!pull) // .. deal with error in `err`
-```
+In order to perform a push or a pull you need to obtain a
+`CDTReplicator` that can to the operation. These are provided by
````
"can to" -> "performs"

jimix replied:
ACK

mkistler commented on doc/coredata.md:
```diff
@@ -99,49 +99,30 @@ NSURL *linkURL = [NSURL URLWithString:databaseURI];
+`CDTReplicator` that can to the operation. These are provided by
+`-replicatorThatPushesToURL:withError:` and
+`-replicatorThatPullsFromURL:withError:`. From this point you can
+assign use the replicator as instructed in [Replication], including
```
"assign use" -> "use"

jimix replied:
ACK

mkistler commented on doc/coredata.md:
````diff
@@ -99,49 +99,30 @@ NSURL *linkURL = [NSURL URLWithString:databaseURI];
-    [NSThread sleepForTimeInterval:1.0f];
 }
 ```
 
````
I thought that there were some requirement on passing conflicts back into the CDTIncrementalStore so that these could be resolved.  Has that changed, or did I misunderstand.  If not, then we need to doc that here.

jimix replied:
This patch is just getting rid of the old/broken push/pull methods.  Conflict resolution will be another PR

mkistler commented on Classes/common/CDTIncrementalStore.h:
```diff
@@ -129,4 +59,34 @@ typedef NS_ENUM(NSInteger, CDTIncrementalStoreErrors) {
+ *  @return a CDTReplicator instance which can be used to start and
+ *  stop the replication itself, or `nil` on error.
+*/
+- (CDTReplicator *)replicatorThatPullsFromURL:(NSURL *)remoteURL withError:(NSError **)error;
```
Since these replicatorThat... methods are instance methods on CDTIncrementalStore, the user will need to first get a reference to the proper CDTIncrementalStore object in order to use them.  I think that is non-obvious, so at a minimum we should give the user some guidance on how to do that, and maybe also consider extending the CDTIncrementalStore interface to make it easier.  E.g. we could offer a class method that takes a peristentStoreCoordinator and storeURL and returns a CDTIncrementalStore object.

jimix replied:
This is covered in `## Accessing your store object`.
Having said that.. the discussion on `-linkReplicators:` needs to go.

jimix committed:
Code Review Comments

This patch contains fixes for:
- Comment typos
- More updates to documentation

Signed-off-by: Jimi Xenidis <jimix@pobox.com>

jimix commented:
Well.. it turns out that the `CDTReplicatorFactory` is single use as well.
At least we have improved the error message :)

mkistler commented:
ACK.  I've reviewed 425a7ed and all looks good to me.

jimix committed:
Allocate a replication factory at the beginning and reuse it

Signed-off-by: Jimi Xenidis <jimix@pobox.com>
